lucene/core/src/java/org/apache/lucene/search/FieldExistsQuery.java

```
@@ -0,0 +1,165 @@
+              iterator = context.reader().getBinaryDocValues(field);
+              break;
+            case SORTED:
+              iterator = context.reader().getSortedDocValues(field);
+              break;
+            case SORTED_NUMERIC:
+              iterator = context.reader().getSortedNumericDocValues(field);
+              break;
+            case SORTED_SET:
+              iterator = context.reader().getSortedSetDocValues(field);
+              break;
+            case NONE:
+            default:
+              throw new AssertionError();
+          }
+        }
+
+        if (iterator == null) {
+          return null;
+        }
+        return new ConstantScoreScorer(this, score(), scoreMode, iterator);
+      }
+
+      @Override
+      public int count(LeafReaderContext context) throws IOException {
+        LeafReader reader = context.reader();
+        FieldInfos fieldInfos = reader.getFieldInfos();
+        FieldInfo fieldInfo = fieldInfos.fieldInfo(field);
+
+        if (fieldInfo == null) {
+          return 0;
+        }
+
+        if (fieldInfo.hasNorms()) { // the field indexes norms
+          // If every field has a value then we can shortcut
+          if (reader.getDocCount(field) == reader.maxDoc()) {
+            return reader.numDocs();
+          }
+
+          return super.count(context);
+        } else if (fieldInfo.getVectorDimension() != 0) { // the field indexes vectors
+          return super.count(context);
+        } else if (fieldInfo.getDocValuesType()
+            != DocValuesType.NONE) { // the field indexes doc values
+          if (reader.hasDeletions() == false) {
+            if (fieldInfo.getPointDimensionCount() > 0) {
+              return reader.getPointValues(field).getDocCount();
+            } else if (fieldInfo.getIndexOptions() != IndexOptions.NONE) {
+              return reader.terms(field).getDocCount();
+            }
+          }
+
+          return super.count(context);
+        }
+
+        return 0;
```

Let's throw an exception here too?

---

Done.